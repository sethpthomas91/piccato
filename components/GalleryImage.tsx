import Router from 'next/router';
import React from 'react';
import styles from '../styles/image.module.css'
import { useState, useEffect } from 'react';
import { getIFPSImageData } from '../handlers/NFTPortHandlers';

export type GalleryImageProps = {
    mint_date: string;
    uri: string;
    metadata_id: string;
}

const GalleryImage: React.FC<{props: GalleryImageProps}> = ({ mint_date, uri, metadata_id } ) => {
  const [ipfsData, setIpfsData] = useState(null);

  useEffect(() => {
    async function fetchData() {
      const newIpfsData = await getIFPSImageData(uri);
      setIpfsData(newIpfsData.data);
    };
    fetchData();
  }, []);

  return (
    <div data-testid="imagePresser" className={styles.container} onClick={() => Router.push('/images/[id]', `/images/${metadata_id}`)}>
      {ipfsData && <img className={styles.galleryImg} src={ipfsData.image}></img>}
    </div>
  );
  }
  
export default GalleryImage;
  