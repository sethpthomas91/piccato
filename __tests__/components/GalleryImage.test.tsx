import renderer from 'react-test-renderer';
import GalleryImage, { GalleryImageProps } from '../../components/GalleryImage';
import { render, fireEvent, screen, renderHook, act } from '@testing-library/react';
import mockRouter from 'next-router-mock';
import { createDynamicRouteParser } from "next-router-mock/dynamic-routes";

jest.mock('next/router', () => require('next-router-mock'));
jest.mock('next/dist/client/router', () => require('next-router-mock'));

describe ('GalleryImage', () => {
  const mockImageProps: GalleryImageProps = {
    mint_date: "date string",
    uri: "uri string",
    metadata_id: "uri",
  }

  it('renders correctly', () => {
    const tree = renderer
      .create(<GalleryImage props={mockImageProps}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  function setupFetchStub(data) {
    return function fetchStub(_url) {
      return new Promise((resolve) => {
        resolve({
          json: () =>
            Promise.resolve({
              data,
            }),
        })
      })
    }
  }

  it('renders a div when useEffect returns a mock image', () => {
    const fakeData = {
      image: "www.testImageLocation.com"
    }
    global.fetch = jest.fn().mockImplementation(setupFetchStub(fakeData))
    render(<GalleryImage props={mockImageProps}/>);
    const imagePresser = screen.getByTestId('imagePresser')
    expect(imagePresser).toBeDefined();
  });

  it('pushes to /images/[id]', () => {
    mockRouter.setCurrentUrl("/");
    mockRouter.useParser(createDynamicRouteParser([
      "/images",
      "/images/[id]",
      "/create",
      '/'
    ]));
    const fakeData = {
      image: "www.testImageLocation.com"
    }
    global.fetch = jest.fn().mockImplementation(setupFetchStub(fakeData))
    
    render(<GalleryImage props={mockImageProps}/>);
    
    act(() => {
      const imagePresser = screen.getByTestId('imagePresser')
      fireEvent.click(imagePresser)});
      expect(mockRouter).toMatchObject({ 
        pathname: "/images/[id]"
      });
    });
})
