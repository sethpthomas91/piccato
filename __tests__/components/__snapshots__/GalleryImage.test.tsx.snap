// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryImage renders correctly 1`] = `
<div
  className="container"
  data-testid="imagePresser"
  onClick={[Function]}
/>
`;
